src/prioritized_pool/step.rs

```diff
@@ -0,0 +1,212 @@
+		for tx in txs {
+			nonces.push(BundleNonce {
+				address: tx.signer(),
+				nonce: tx.nonce(),
+				optional: bundle.is_optional(tx.tx_hash()),
+			});
+		}
+		// for each address we keep lowest nonce
+		nonces
+			.sort_by(|a, b| a.address.cmp(&b.address).then(a.nonce.cmp(&b.nonce)));
+		nonces.dedup_by_key(|n| n.address);
+		Self {
+			bundle,
+			nonces,
+			phantom: PhantomData,
+		}
+	}
+}
+
+impl<B, P> PrioritizedOrderpoolOrder for BundleWithNonces<B, P>
+where
+	B: Bundle<P>,
+	P: Platform,
+{
+	type ID = B256;
+
+	fn id(&self) -> Self::ID {
+		self.bundle.hash()
+	}
+
+	fn nonces(&self) -> Vec<BundleNonce> {
+		self.nonces.clone()
+	}
+}
+
+impl<P: Platform> PrioritizedOrderpoolNonceSource for Checkpoint<P> {
+	type NonceError = reth::errors::ProviderError;
+
+	fn nonce(&self, address: &Address) -> Result<u64, Self::NonceError> {
+		Ok(
+			self
+				.basic_ref(*address)?
+				.map(|acc| acc.nonce)
+				.unwrap_or_default(),
+		)
+	}
+}
+
+pub struct EffectiveGasPriceOrdering<B, P> {
+	effective_gas_price: u128,
+	marker: PhantomData<(B, P)>,
+}
+
+impl<B, P> Clone for EffectiveGasPriceOrdering<B, P> {
+	fn clone(&self) -> Self {
+		Self {
+			effective_gas_price: self.effective_gas_price,
+			marker: PhantomData,
+		}
+	}
+}
+
+impl<B, P> PartialOrd for EffectiveGasPriceOrdering<B, P> {
+	fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+		Some(self.cmp(other))
+	}
+}
+
+impl<B, P> Ord for EffectiveGasPriceOrdering<B, P> {
+	fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+		self.effective_gas_price.cmp(&other.effective_gas_price)
+	}
+}
+
+impl<B, P> PartialEq for EffectiveGasPriceOrdering<B, P> {
+	fn eq(&self, other: &Self) -> bool {
+		self.effective_gas_price == other.effective_gas_price
+	}
+}
+impl<B, P> Eq for EffectiveGasPriceOrdering<B, P> {}
+
+impl<B: Send + Sync + Bundle<P>, P: Send + Sync + Platform>
+	PrioritizedOrderpoolPriority for EffectiveGasPriceOrdering<B, P>
+{
+	type Order = BundleWithNonces<B, P>;
+
+	fn new(order: &Self::Order) -> Self {
+		let mut total_gas_limit = 0u128;
+		let mut accumulated_weighted_fee = 0u128;
+		for tx in order.bundle.transactions() {
+			if order.bundle.is_optional(tx.tx_hash()) {
+				continue;
+			}
+			let max_fee = tx.max_fee_per_gas();
+			let gas_limit = u128::from(tx.gas_limit());
+			accumulated_weighted_fee += max_fee * gas_limit;
+			total_gas_limit += gas_limit;
+		}
+		let effective_gas_price = if total_gas_limit != 0 {
+			accumulated_weighted_fee / total_gas_limit
+		} else {
+			0
+		};
+		Self {
+			effective_gas_price,
+			marker: PhantomData,
+		}
+	}
+}
+
+type PrioritizedOrderpoolForPlatform<Priority, P: Platform> =
+	PrioritizedOrderpool<Priority, BundleWithNonces<P::Bundle, P>>;
+
+pub struct AppendOrdersByPriority<P, Pending, Priority>
+where
+	P: Platform,
+	Pending: PendingOrdersSource<P>,
+	Priority:
+		PrioritizedOrderpoolPriority<Order = BundleWithNonces<P::Bundle, P>>,
+{
+	pending: Pending,
+	prioritized_orderpool:
```

Comment:
I would try to not make assumptions about the execution of the pipeline that is using this step. You might have many concurrent pipelines running a step and if they all share the same orderpool we might get undefined behavior.

If I'm understanding the intention here correctly (and the `&mut self` accessors on `PrioritizedOrderpool` I would recommend one of the two following paths forward:

1. Either the step consumes the `PrioritizedOrderpool` by value and `PrioritizedOrderpool` maintains some form of a `mpsc::Receiver` for incoming orders,
2. Or use something like [im](https://docs.rs/im/latest/im/) that gives immutable data structures and each instance of the step operates on a snapshot/clone of it.